### PR TITLE
Add command for resetting diff hunks

### LINKED ISSUE #4974: Add command to reset diff hunks
Suggested to me by @Byron

It would be nice to have a command/keybinding to reset hunks (like `git reset --hard` but more fine grained).
This functionality is diff provider independent and can be implemented by just removing the `after` lines in the current document and replacing the with the `before` lines in the diff base.

I would implement this as a type able command (maybe `:reset` ) that resets the hunk at the current cursor. Ideally there would be a way to quickly place multi-cursors inside hunks so that multiple hunks can be selected at once.

### PR DESCRIPTION
**Pull request by pascalkuthe**
Closes #4974

Adds the `:reset-diff-change` command  (aliases `:diffg` and `:diffget`) that resets a hunk back
to its original text. 

This change requires exposing the `diff_base` (the file that a diff was computed against). To accommodate that I slightly refactored the terminology the `DiffHandle` and renamed the `FileHunks` struct to `Diff` as it doesn't just contains hunks anymore.

### CODE REVIEW

File: helix-term/src/commands/typed.rs [hunk 1]
```diff
@@ -1822,6 +1822,60 @@ fn run_shell_command(
...
+    if event != PromptEvent::Validate {
+        return Ok(());
+    }
+    ensure!(args.is_empty(), ":reset-change takes no arguments");
+
+    let editor = &mut cx.editor;
+    let scrolloff = editor.config().scrolloff;
+
+    let (view, doc) = current!(editor);
+    if let Some(handle) = doc.diff_handle() {
```
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:1839]: This isn't a super big block but the indentation here could be reduced:

```rs
let handle = match doc.diff_handle() {
    Some(handle) => handle,
    None => bail!("Diff is not available in the current buffer"),
};

// ...
```

and the same for the next `if let Some`

Comment (pascalkuthe) [reply helix-term/src/commands/typed.rs]: that does look cleaner, thanks! This is actually an ideal case for `let ... else` but sadly we haven't reached a MSRV of 1.65 yet. Added a `TODO` for that

**Comment by trink** [comment]
The diagnostics get confused by the reset. It just pushed the diagnostic down and I have to save the file to clear it. Although this isn't specific to reset, it is just easier to demonstrate here.

Peek 2023-01-30 11-39

**Comment by pascalkuthe** [comment]
> The diagnostics get confused by the reset. It just pushed the diagnostic down and I have to save the file to clear it. Although this isn't specific to reset, it is just easier to demonstrate here.
> 
> Peek 2023-01-30 11-39 [ Peek 2023-01-30 11-39 ](https://user-images.githubusercontent.com/795988/215578823-d608aacb-72e0-42d6-b631-a7db017f799a.gif) [ ](https://user-images.githubusercontent.com/795988/215578823-d608aacb-72e0-42d6-b631-a7db017f799a.gif)

That's just #701 and happens for any edit not just the `:reset-change` command here (also the case on master) so unrelated to this PR.

File: helix-term/src/commands/typed.rs [hunk 2]
```diff
@@ -2337,6 +2395,13 @@ pub const TYPABLE_COMMAND_LIST: &[TypableCommand] = &[
             fun: run_shell_command,
             completer: Some(completers::directory),
         },
+       TypableCommand {
+            name: "reset-change",
+            aliases: &["reset"],
+            doc: "Undo the change at cursor position",
+            fun: reset_change,
+            completer: None,
+        },
```
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:2404]: I'm not sure where to put this but I think either the name or description (or both) should say "VCS". Otherwise this might be confused with https://github.com/helix-editor/helix/issues/3364

Comment (pascalkuthe) [reply helix-term/src/commands/typed.rs]: The reason I shy away from using `vcs` in diff related functionality is that these commands would also be used for the diffmode where we might not actually be comparing against a VCS file but rather just a different buffer/file on disk. 

I would have used something like `undo-selection` for #3364. So any command that goes back in undo history would contain `undo` and `reset` would be for the diff. I can see that the names could be potentially confusing tough. This is sort of a bikeshed issue.


For now I just changed the description so that it uses `reset` instead of `undo` (to keep terminology straight) and explicitly refer to a`diff change`. Do you think that is enough or would you rather see the name changed.
Some options I came up wiith:

* `:reset-change`: What I choose initially for this PR.
* `:reset-diff-change`: Probably most concise but a bit wordy
* `:reset-diff`: Sounds like the entire diff is reset
* `:diffget` (alias `:diffg`): Name stolen from nvim. The nice thing is that it has a cousing `:diffput` to reset the other side of the diff if we are comparing to editable files (not currently implemented)

Comment (pascalkuthe) [reply helix-term/src/commands/typed.rs]: Thanks for your input. I renamed to command to `:reset-diff-change`.
I decided to add `:diffget` and `:diffg` as aliases. These names are quite well know from other editors, and I don't think they are not easily confused with something else.
I personally don't bind this to a keybinding so I also appreciate being able to type `:diffg` as a convenience (similar to how we have `:w` as an alias for `:write`).

I also think that it will make it clearer that this command is closely linked to a future command that changes the other side of the disk (which would alias to `:diffput`)

**Review by the-mikedavis — DISMISSED** [review]
